---
title: "ulrb-vignet"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{ulrb-vignet}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Unsupervised Learning Based Definition Of Microbial Rare Biosphere

The R package ulrb solves the problem of the definition of rarity in molecular microbiology by replacing human decision with a unsupervised machine learning algorithm (kmedoids). This algorithm works for any type of microbiome data, provided there is an abundance score for each taxonomic unit. For validation of this method to several kinds of molecular data and environments, please see REF.

In this vignete we will illustrate a full analysis of the rare biosphere with a small, publically available dataset, the Norwegian Young Sea Ice Expedition, 2016 (N-ICE). The molecular data will be from amplicon sequencing of 16S rRNA gene (for sequencing details, see Sousa et al., 2019), raw sequences are available in the European Nucleotide Archive (ENA) under accession number #####, and they were processed into OTUs by MGnify (for details on MGnify bioinformatics processing see Mitchel et al.,).

The N-ICE dataset consists of 9 samples, collected during the winter to spring transition of 2015 by fixing the vessel on drifting ice. Samples were collected at various depths (from 5m to 250m) and the ice drifted across different regions.

For this vignete we are using the OTU table downloaded from the link https://www.ebi.ac.uk/metagenomics/studies/MGYS00001922#analysis in 06-01-2023.

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(ulrb)
library(dplyr)
library(tidyr)
```

### Brief note on nomenclature

We use the term "taxonomic units" to refer to the highest resolution, distinct unit identified. This can be species, OTU, ASV...

We use the term "abundance table" to refer to the table with, at least, the abundance score for each taxonomic unit. Any additional data can be added, this can be an OTU table, ASV table... 

We also use the term "scale" or "scaling", which can also appear in the literature as "normalization".

## Pre-processing of data prior to kmedoids algorithm

The package ulrb will require that the abundance table is in tidy format, with a column for the abundance score used and another column for samples. Any additional columns are allowed.

**Regarding the abundance score**, we are not scaling the abundance socre in this vignete, because the kmedoids algorithm will be applied "per sample" to the values of the abundance score. This means that any scalling, e.g. relative abundance, clr... will not have an effect on the algorithm, because the distance values between observations will remain the same. However, if the algorithm was applied to more variables (not only abundance), then scaling could make a difference. Thus, we consider that is best to use absolute values in this tutorial, to avoid misleading the reader into thinking that some scaling has to be done. 

Notwistanding, for publication purposes and for the purposes of other data processing steps, it might make sense to apply some sort of scaling, as long as the number of observations, i.e. the number of taxonomic units, is the same, then the distance between them will be the same.

If the user decides to apply any quality filtering to the abundance table, like rarefaction, it should be done *before* the kmedoids algorithm.

In summary, the algorithm will be applied to your final, clean abundance table, with any abundance score.

### Step 0 - Load and clean abundance table

```{r}
# Load raw OTU table from N-ICE
load("../data/nice_raw.rda")

# Change name of first column
nice_clean <- rename(nice_raw, Taxonomy = "X.SampleID")

# Select 16S rRNA amplicon sequencing samples
selected_samples <- c("ERR2044662", "ERR2044663", "ERR2044664","ERR2044665", "ERR2044666", "ERR2044667","ERR2044668", "ERR2044669", "ERR2044670")

# Add a column with taxonomic units ID (OTU in this case)
nice_clean <- mutate(nice_clean, OTU = paste0("OTU_", row_number()))

# Select relevant collumns
nice_clean <- select(nice_clean, selected_samples, OTU, Taxonomy)

# Separate Taxonomy column into each taxonomic level
nice_clean <- separate(nice_clean, Taxonomy, c("Domain","Kingdom","Phylum","Class","Order","Family","Genus","Species"),sep=";")

# Remove Kingdom column, because it is not used for prokaryotes
nice_clean <- select(nice_clean, -Kingdom)

# Remove eukaryotes
nice_clean <- filter(nice_clean, Domain != "sk__Eukaryota")

# Remove unclassified OTUs at phylum level
nice_clean <- filter(nice_clean, !is.na(Phylum))

# Simplify name
nice <- nice_clean

# Check if everything looks normal
head(nice)
```

### Step 1 - Transform abundance table into tidy format

Now that we have a cleaned OTU table, we can transform it to tidy format. You can do this part in any order and according to your own data format. We made a function to help transform the two most common data formats for abundance tables:
1) taxonomic units in rows, samples in collumns;
2) taxonomic units in collumns, samples in rows.

In this example, we have a data.frame with taxonomic units as rows and samples in various collumns, so this is not a tidy format. The transformation we need to do is to collapse all columns with samples into a single "Abundance" and "Samples" collumn, so that each variable corresponds to a single collumn, i.e. tidy.

For this we just need to apply the prepare_tidy_data() function to the cleaned data. Note that you need to specify in which format your abundance table is, in this case, samples are in different collumns, so we set the argument samples_in to "cols"; is it was the opposite situation, with samples in rows, we would set samples_in to "rows".

This function is optional and if you use it you really should verify if it is transforming the data correctly, because your own data might be different from our standard examples.

```{r}
nice_tidy <- prepare_tidy_data(nice, sample_names = selected_samples, samples_in = "cols")

head(nice_tidy)
```


