<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="ulrb">
<title>Alternative classifications with ulrb • ulrb</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Alternative classifications with ulrb">
<meta property="og:description" content="ulrb">
<meta property="og:image" content="https://pascoalf.github.io/ulrb/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">ulrb</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/Glossary.html">Glossary</a>
    <a class="dropdown-item" href="../articles/eco-analysis.html">Integration of ulrb in a simple microbial ecology workflow</a>
    <a class="dropdown-item" href="../articles/explore-classifications.html">Alternative classifications with ulrb</a>
    <a class="dropdown-item" href="../articles/ulrb-vignet.html">Tutorial to define rare biosphere with ulrb</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/pascoalf/ulrb/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Alternative classifications with ulrb</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/pascoalf/ulrb/blob/HEAD/vignettes/explore-classifications.Rmd" class="external-link"><code>vignettes/explore-classifications.Rmd</code></a></small>
      <div class="d-none name"><code>explore-classifications.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://pascoalf.github.io/ulrb/">ulrb</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://svn.r-project.org/R-packages/trunk/cluster/" class="external-link">cluster</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org" class="external-link">tidyr</a></span><span class="op">)</span></span>
<span><span class="co"># a vector with some colors</span></span>
<span><span class="va">qualitative_colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#E69F00"</span>, <span class="st">"#56B4E9"</span>, <span class="st">"#009E73"</span>, <span class="st">"#F0E442"</span>, <span class="st">"#0072B2"</span>, <span class="st">"#D55E00"</span>, <span class="st">"#CC79A7"</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="explore-alternative-classifications">Explore alternative classifications<a class="anchor" aria-label="anchor" href="#explore-alternative-classifications"></a>
</h2>
<p>By default, we have been considering that the microbial communities
are divided into “rare”, “undetermined” and “abundant”. This division
implies that the partition around medoids algorithm is considering
<strong>k=3</strong>, <em>i.e.</em> there are three clusters. However,
there are at least three situations where you might want to change the
number of clusters:</p>
<ol style="list-style-type: decimal">
<li>The clusters obtained are non-sense;</li>
<li>The biological/ecological questions requires another division;</li>
<li>You want a fully automated classification.</li>
</ol>
<p>There might be other situations, of course. The bottom line is that,
depending on the context of your specific research and data, you might
just want a different number of classifications. We propose 3 and we
think that 3 different classifications work well for most situations,
but this is not set in stone.</p>
<p>In this tutorial we will show how <code><a href="../reference/define_rb.html">define_rb()</a></code> can be
used for different classifications and we will also try and go a little
bit deeper on whats behind the actual function.</p>
</div>
<div class="section level2">
<h2 id="index">Index<a class="anchor" aria-label="anchor" href="#index"></a>
</h2>
<ol style="list-style-type: decimal">
<li><p>Classical example;</p></li>
<li><p>Apply 2 classifications: Rare vs Abundant;</p></li>
<li><p>Apply more complicated classification, k&gt;3;</p></li>
<li><p>Why k = 1 is non-sense;</p></li>
<li><p>What is the maximum number of k and why?;</p></li>
<li><p>Approaches to evaluate k;</p></li>
<li><p>Set k automatically.</p></li>
</ol>
<div class="section level3">
<h3 id="classical-example">Classical example<a class="anchor" aria-label="anchor" href="#classical-example"></a>
</h3>
<p>The classical example will simply use the default arguments with the
function <code><a href="../reference/define_rb.html">define_rb()</a></code>.</p>
<div class="section level4">
<h4 id="prepare-data">Prepare data<a class="anchor" aria-label="anchor" href="#prepare-data"></a>
</h4>
<p>Just like in the tutorial <code><a href="../articles/ulrb-vignet.html">vignette("ulrb-vignet")</a></code>, we
have to import, clean and tidy the dataset before using the
<code><a href="../reference/define_rb.html">define_rb()</a></code> function.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load raw OTU table from N-ICE</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="st">"../data/nice_raw.rda"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Change name of first column</span></span>
<span><span class="va">nice_clean</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span><span class="va">nice_raw</span>, Taxonomy <span class="op">=</span> <span class="st">"X.SampleID"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Select 16S rRNA amplicon sequencing samples</span></span>
<span><span class="va">selected_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ERR2044662"</span>, <span class="st">"ERR2044663"</span>, <span class="st">"ERR2044664"</span>,</span>
<span>                      <span class="st">"ERR2044665"</span>, <span class="st">"ERR2044666"</span>, <span class="st">"ERR2044667"</span>,</span>
<span>                      <span class="st">"ERR2044668"</span>, <span class="st">"ERR2044669"</span>, <span class="st">"ERR2044670"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add a column with phylogenetic units ID (OTU in this case)</span></span>
<span><span class="va">nice_clean</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span><span class="va">nice_clean</span>, OTU <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"OTU_"</span>, <span class="fu"><a href="https://dplyr.tidyverse.org/reference/row_number.html" class="external-link">row_number</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Select relevant collumns</span></span>
<span><span class="va">nice_clean</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">nice_clean</span>, <span class="va">selected_samples</span>, <span class="va">OTU</span>, <span class="va">Taxonomy</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Using an external vector in selections was deprecated in tidyselect 1.1.0.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Please use `all_of()` or `any_of()` instead.</span></span>
<span><span class="co">#&gt;   # Was:</span></span>
<span><span class="co">#&gt;   data %&gt;% select(selected_samples)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   # Now:</span></span>
<span><span class="co">#&gt;   data %&gt;% select(all_of(selected_samples))</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; See &lt;https://tidyselect.r-lib.org/reference/faq-external-vector.html&gt;.</span></span>
<span><span class="co">#&gt; <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">generated.</span></span></span>
<span></span>
<span><span class="co"># Separate Taxonomy column into each taxonomic level</span></span>
<span><span class="va">nice_clean</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tidyr.tidyverse.org/reference/separate.html" class="external-link">separate</a></span><span class="op">(</span><span class="va">nice_clean</span>, </span>
<span>                       <span class="va">Taxonomy</span>,</span>
<span>                       <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Domain"</span>,<span class="st">"Kingdom"</span>,<span class="st">"Phylum"</span>,</span>
<span>                         <span class="st">"Class"</span>,<span class="st">"Order"</span>,<span class="st">"Family"</span>,</span>
<span>                         <span class="st">"Genus"</span>,<span class="st">"Species"</span><span class="op">)</span>, </span>
<span>                       sep<span class="op">=</span><span class="st">";"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Expected 8 pieces. Missing pieces filled with `NA` in 912 rows [1, 2, 4, 5, 6,</span></span>
<span><span class="co">#&gt; 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, ...].</span></span>
<span></span>
<span><span class="co"># Remove Kingdom column, because it is not used for prokaryotes</span></span>
<span><span class="va">nice_clean</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">nice_clean</span>, <span class="op">-</span><span class="va">Kingdom</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Remove eukaryotes</span></span>
<span><span class="va">nice_clean</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">nice_clean</span>, <span class="va">Domain</span> <span class="op">!=</span> <span class="st">"sk__Eukaryota"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Remove unclassified OTUs at phylum level</span></span>
<span><span class="va">nice_clean</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">nice_clean</span>, <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">Phylum</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Simplify name</span></span>
<span><span class="va">nice</span> <span class="op">&lt;-</span> <span class="va">nice_clean</span></span>
<span></span>
<span><span class="co"># Tidy data</span></span>
<span><span class="va">nice_tidy</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepare_tidy_data.html">prepare_tidy_data</a></span><span class="op">(</span><span class="va">nice</span>, </span>
<span>                               sample_names <span class="op">=</span> <span class="va">selected_samples</span>, </span>
<span>                               samples_in <span class="op">=</span> <span class="st">"cols"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="default-is-k-3">Default is k = 3<a class="anchor" aria-label="anchor" href="#default-is-k-3"></a>
</h4>
<p>The default settings can be interpreted as the division between
“rare”, “undetermined” and “abundant” species.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rb_default</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/define_rb.html">define_rb</a></span><span class="op">(</span><span class="va">nice_tidy</span><span class="op">)</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(Sample, Level)`</span></span></code></pre></div>
<p>We can see the meaning of this division with the standard Rank
Abundance Curve (RAC).</p>
<p>Let’s start with a single sample:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_ulrb_clustering.html">plot_ulrb_clustering</a></span><span class="op">(</span><span class="va">rb_default</span>, </span>
<span>                     sample_id <span class="op">=</span> <span class="st">"ERR2044662"</span>,</span>
<span>                     taxa_col <span class="op">=</span> <span class="st">"OTU"</span>,</span>
<span>                     log_scaled <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="explore-classifications_files/figure-html/unnamed-chunk-4-1.png" width="480"></p>
<p>To look at all samples at once, we use a centrality metric. Note that
for a low number of samples you can chose to use a grid, or plot all
samples; but this wont work for a high number of samples. With a
centrality metric, however, we are able to see any number of samples in
a single plot.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_ulrb_clustering.html">plot_ulrb_clustering</a></span><span class="op">(</span><span class="va">rb_default</span>, </span>
<span>                     taxa_col <span class="op">=</span> <span class="st">"OTU"</span>,</span>
<span>                     log_scaled <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     plot_all <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 197 rows containing missing values (`geom_segment()`).</span></span></code></pre></div>
<p><img src="explore-classifications_files/figure-html/unnamed-chunk-5-1.png" width="768"></p>
</div>
</div>
<div class="section level3">
<h3 id="apply-2-classifications-rare-vs-abundant">Apply 2 classifications: Rare vs Abundant<a class="anchor" aria-label="anchor" href="#apply-2-classifications-rare-vs-abundant"></a>
</h3>
<p>We can do exactly the same thing we did before, but with only “rare”
and “abundant” classifications, to do that we set k = 2. Note that we do
not explicitly set k = 2 in the <code><a href="../reference/define_rb.html">define_rb()</a></code> function,
instead, we specify our <strong>classification vector</strong>.</p>
<p>The classification vector can be anything, but in the context of k =
2, our meaning is that we have “rare” and “abundant” species; but you
could call them “class1” and “class2”, or whatever you want. Behind the
scenes, the function <code><a href="../reference/define_rb.html">define_rb()</a></code> will calculate the size of
your classification vector and from there estimate k. So, if you have a
classification vector with “rare” and “abundant”, it will have length 2,
so k = 2; Likewise, the default parameter is set to
classification_vector = c(“Rare”, “Undetermined”, “Abundant”), so length
= 3 and k = 3.</p>
<p>To apply this new classification scheme we change the
<code>classification_vector</code> argument:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rb_k2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/define_rb.html">define_rb</a></span><span class="op">(</span><span class="va">nice_tidy</span>, classification_vector <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Rare"</span>, <span class="st">"Abundant"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(Sample, Level)`</span></span></code></pre></div>
<p>And then we can see the clustering result just like before.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_ulrb_clustering.html">plot_ulrb_clustering</a></span><span class="op">(</span><span class="va">rb_k2</span>,</span>
<span>                     taxa_col <span class="op">=</span> <span class="st">"OTU"</span>,</span>
<span>                     plot_all <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                     log_scaled <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                     colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#009E73"</span>, <span class="st">"#F0E442"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 177 rows containing missing values (`geom_segment()`).</span></span></code></pre></div>
<p><img src="explore-classifications_files/figure-html/unnamed-chunk-7-1.png" width="480"></p>
<p>And we can compare the two options directly:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">gridExtra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/gridExtra/man/arrangeGrob.html" class="external-link">grid.arrange</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/plot_ulrb_clustering.html">plot_ulrb_clustering</a></span><span class="op">(</span><span class="va">rb_default</span>, </span>
<span>                     taxa_col <span class="op">=</span> <span class="st">"OTU"</span>,</span>
<span>                     log_scaled <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     plot_all <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="../reference/plot_ulrb_clustering.html">plot_ulrb_clustering</a></span><span class="op">(</span><span class="va">rb_k2</span>,</span>
<span>                     taxa_col <span class="op">=</span> <span class="st">"OTU"</span>,</span>
<span>                     plot_all <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                     log_scaled <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                     colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#009E73"</span>, <span class="st">"#F0E442"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  nrow <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 197 rows containing missing values (`geom_segment()`).</span></span>
<span><span class="co">#&gt; Warning: Removed 177 rows containing missing values (`geom_segment()`).</span></span></code></pre></div>
<p><img src="explore-classifications_files/figure-html/unnamed-chunk-8-1.png" width="1440"></p>
<p>We do not recommend using k = 2, because it implies that there is a
hard distinction between rare and abundant taxa. We consider that there
should always be an intermediate or undetermined group, otherwise, we
will always have phylogenetic units with very similar abundance scores,
but opposing classifications, which is misleading. For more details on
this reasoning, see Pascoal et al., 2023 (manuscript in
preparation).</p>
</div>
<div class="section level3">
<h3 id="apply-more-complicated-classification-k3">Apply more complicated classification, k&gt;3<a class="anchor" aria-label="anchor" href="#apply-more-complicated-classification-k3"></a>
</h3>
<p>Lets suppose that we want to distinguish our microbial community into
the following groups:</p>
<p>(<em>e.g.</em> A) 1) very rare; 2) rare; 3) abundant; 4) very
abundant.</p>
<p>Or:</p>
<p>(<em>e.g.</em> B) 1) very rare; 2) rare; 4) undetermined; 5)
abundant; 6) very abundant.</p>
<p>Or even more…</p>
<p>Hopefully, you get the idea, the point is that we just have to change
the classification vector.</p>
<p>Lets see how we would do <em>e.g.</em> A. First we select a new
classification vector, which will have 4 different classifications, thus
(implicitly) k = 4.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#</span></span>
<span><span class="va">rb_k4</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/define_rb.html">define_rb</a></span><span class="op">(</span><span class="va">nice_tidy</span>, </span>
<span>                   classification_vector <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"very rare"</span>, <span class="st">"rare"</span>, <span class="st">"abundant"</span>, <span class="st">"very abundant"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(Sample, Level)`</span></span>
<span><span class="co">#</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># One sample as example</span></span>
<span><span class="fu"><a href="../reference/plot_ulrb.html">plot_ulrb</a></span><span class="op">(</span><span class="va">rb_k4</span>,</span>
<span>          sample_id <span class="op">=</span> <span class="st">"ERR2044662"</span>, </span>
<span>          taxa_col <span class="op">=</span> <span class="st">"OTU"</span>,</span>
<span>          colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#009E73"</span>, <span class="st">"#F0E442"</span>, <span class="st">"grey"</span>,<span class="st">"#CC79A7"</span><span class="op">)</span>,</span>
<span>          log_scaled <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="explore-classifications_files/figure-html/unnamed-chunk-10-1.png" width="1440"></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># all samples</span></span>
<span><span class="fu"><a href="../reference/plot_ulrb.html">plot_ulrb</a></span><span class="op">(</span><span class="va">rb_k4</span>,</span>
<span>          taxa_col <span class="op">=</span> <span class="st">"OTU"</span>,</span>
<span>          colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#009E73"</span>, <span class="st">"#F0E442"</span>, <span class="st">"grey"</span>,<span class="st">"#CC79A7"</span><span class="op">)</span>,</span>
<span>          log_scaled <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>          plot_all <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 208 rows containing missing values (`geom_segment()`).</span></span>
<span><span class="co">#&gt; Removed 208 rows containing missing values (`geom_segment()`).</span></span></code></pre></div>
<p><img src="explore-classifications_files/figure-html/unnamed-chunk-11-1.png" width="1440"></p>
<p>Note that this is a very acceptable interpretation of the data!</p>
<p>Hopefully you already have the intuition to change classifications as
you wish. Let’s do the <em>e.g.</em> B, which has k = 5, because it
includes 5 different classifications.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#</span></span>
<span><span class="va">rb_k5</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/define_rb.html">define_rb</a></span><span class="op">(</span><span class="va">nice_tidy</span>, </span>
<span>                   classification_vector <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"very rare"</span>, <span class="st">"rare"</span>, <span class="st">"undetermined"</span>, <span class="st">"abundant"</span>, <span class="st">"very abundant"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(Sample, Level)`</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># One sample as example</span></span>
<span><span class="fu"><a href="../reference/plot_ulrb.html">plot_ulrb</a></span><span class="op">(</span><span class="va">rb_k5</span>,</span>
<span>          sample_id <span class="op">=</span> <span class="st">"ERR2044662"</span>, </span>
<span>          taxa_col <span class="op">=</span> <span class="st">"OTU"</span>,</span>
<span>          colors <span class="op">=</span> <span class="va">qualitative_colors</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>,</span>
<span>          log_scaled <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="explore-classifications_files/figure-html/unnamed-chunk-13-1.png" width="1440"></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># All samples</span></span>
<span><span class="fu"><a href="../reference/plot_ulrb.html">plot_ulrb</a></span><span class="op">(</span><span class="va">rb_k5</span>,</span>
<span>          taxa_col <span class="op">=</span> <span class="st">"OTU"</span>,</span>
<span>          colors <span class="op">=</span> <span class="va">qualitative_colors</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>,</span>
<span>          log_scaled <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>          plot_all <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 213 rows containing missing values (`geom_segment()`).</span></span>
<span><span class="co">#&gt; Removed 213 rows containing missing values (`geom_segment()`).</span></span></code></pre></div>
<p><img src="explore-classifications_files/figure-html/unnamed-chunk-14-1.png" width="1440"></p>
</div>
<div class="section level3">
<h3 id="why-k-1-is-non-sense">Why k = 1 is non-sense<a class="anchor" aria-label="anchor" href="#why-k-1-is-non-sense"></a>
</h3>
<p>Some values of k are (mathematically) possible, but non-sense.
Consider, for example, <strong>k = 1</strong>. In this situation, it is
perfectly possible from a mathematical point of view, <em>i.e</em> you
simply cluster all observations into a single classification. However,
this is meaningless and gives no extra information on your community.
Additionally, from the point of view of the rare biosphere, it is
important to keep in mind that one phylogenetic unit can only be
considered “rare” <strong>relatively</strong> to the other phylogenetic
units within the same community. Thus, in this context, <strong>k = 1 is
self-contradictory</strong>. Despite this, do keep in mind that we are
making the assumption that there is such thing as a “rare biosphere”.
This naive assumption caries another assumption within itself, which is
that phylogenetic units have sufficiently different abundance relative
to one another to form distinct clusters. In fact, I don’t think anyone
as found an environmental microbial community constituted of
phylogenetic units with the exactly (or approximately) same abundance
score. If that was the case, then we could not say that a species was
neither rare or abundant, because any classification would be
self-contradictory.</p>
<p>We can do the same thing as before, setting the classification vector
to just “rare” and you’ll see how it is possible, but meaningless.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#</span></span>
<span><span class="va">rb_k1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/define_rb.html">define_rb</a></span><span class="op">(</span><span class="va">nice_tidy</span>, classification_vector <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rare"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(Sample, Level)`</span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_ulrb_clustering.html">plot_ulrb_clustering</a></span><span class="op">(</span><span class="va">rb_k1</span>, </span>
<span>                     taxa_col <span class="op">=</span> <span class="st">"OTU"</span>, </span>
<span>                     colors <span class="op">=</span> <span class="st">"green4"</span>, </span>
<span>                     plot_all <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                     log_scaled <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 168 rows containing missing values (`geom_segment()`).</span></span></code></pre></div>
<p><img src="explore-classifications_files/figure-html/unnamed-chunk-16-1.png" width="480">
## What is the maximum value of k and why?</p>
<p>What about the maximum value of k? This is set by the limitations
inherent to algorithm itself, which in this case is partition around
medoids (k-medoids). The maximum number of groups on which we divide
observations must be less than the total observations. Meaning that at
the maximum limit, each possible observation is its own group. This
means that it could be possible to have a maximum value of k equal to
the number of different phylogenetic units, <em>i.e.</em> each
phylogenetic unit has its own classification. However, if you try to do
that, you will get an error. This is because several phylogenetic units
have exactly the same abundance, specially in the rare biosphere. Thus,
those phylogenetic units can never belong to different classifications
(if we are separating phylogenetic units by their abundance, then
phylogenetic units with the same abundance must belong to the same
group). This means that the maximum value of k will be equal to number
of <strong>different</strong> abundance scores, which will be lower than
the number of total observations.</p>
<p>If you have several samples, then the maximum value of k will differ
for each sample (even if you apply normalization). If you would apply
the same maximum k for all samples at once, then you would have to
calculate the maximum k of all samples, and then select the minimum of
those, which will be the one working for all.</p>
<p>Just like with k = 1, the maximum value of k is meaningless. Even if
you do it, which you mathematically can, it will not give you any
additional information.</p>
<p>Lets illustrate the maximum k of one sample:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rb_sample1</span> <span class="op">&lt;-</span> <span class="va">nice_tidy</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">Sample</span> <span class="op">==</span> <span class="st">"ERR2044662"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Calculate maximum k</span></span>
<span><span class="va">max_k_sample1</span> <span class="op">&lt;-</span> <span class="va">rb_sample1</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">Abundance</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#</span></span>
<span><span class="va">max_k_sample1</span></span>
<span><span class="co">#&gt; [1] 71</span></span>
<span><span class="co"># Improvise a classification vector for maximum k</span></span>
<span><span class="co"># that is just any vector with the same length</span></span>
<span><span class="va">classification_vector_max_k_sample1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">max_k_sample1</span><span class="op">)</span></span>
<span><span class="co">#</span></span>
<span><span class="va">rb_sample1_max_k</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/define_rb.html">define_rb</a></span><span class="op">(</span><span class="va">rb_sample1</span>,</span>
<span>            classification_vector <span class="op">=</span> <span class="va">classification_vector_max_k_sample1</span><span class="op">)</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(Sample, Level)`</span></span>
<span><span class="co">#&gt; Warning in define_rb(rb_sample1, classification_vector =</span></span>
<span><span class="co">#&gt; classification_vector_max_k_sample1): 25 samples got a bad Silhouette score.</span></span>
<span><span class="co">#&gt; Consider changing the number of classifications.</span></span>
<span><span class="co">#&gt; If half the observations within a classification are below 0.5 Silhouette score, we consider that the clustering was 'Bad'.</span></span>
<span><span class="co">#&gt; Check 'Evaluation' collumn for more details.</span></span>
<span><span class="co">#</span></span>
<span><span class="va">rb_sample1_max_k</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">OTU</span>, <span class="va">Classification</span>, <span class="va">Abundance</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt; Adding missing grouping variables: `Sample`</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 10 × 4</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Groups:   Sample, Classification [3]</span></span></span>
<span><span class="co">#&gt;    Sample     OTU     Classification Abundance</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> ERR2044662 OTU_2   44                   165</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> ERR2044662 OTU_5   58                   541</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> ERR2044662 OTU_6   8                      8</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> ERR2044662 OTU_29  8                      8</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> ERR2044662 OTU_30  8                      8</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> ERR2044662 OTU_244 8                      8</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> ERR2044662 OTU_260 8                      8</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> ERR2044662 OTU_313 8                      8</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> ERR2044662 OTU_329 8                      8</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> ERR2044662 OTU_380 8                      8</span></span></code></pre></div>
<p>I’m not going to plot this figure, because I would need 71 different
colors! There really is no point, this was just to show you that it is
possible. With some data wrangling you could easily do the same for all
samples, or any number of samples. But this is really not something that
you should ever do in the context of rare biosphere studies.</p>
</div>
<div class="section level3">
<h3 id="approaches-to-evaluate-k">Approaches to evaluate k<a class="anchor" aria-label="anchor" href="#approaches-to-evaluate-k"></a>
</h3>
<p>What if we want to compare all possible values of k? Or, if not all
of them, just a few candidates?</p>
<p>To compare values of k, there are a few useful metrics that are
implemented in ulrb package:</p>
<ul>
<li>Average Silhouette score (cluster density);</li>
<li>Davies-Bouldin index (cluster separation);</li>
<li>And Calinsky-Harabasz index (cluster definition).</li>
</ul>
<p>Each of these metrics evaluates different aspects of one clustering
result; thus, we can calculate them for the clustering obtained by each
value of k. By comparing the values obtained for each k, you can select
the k that got the best score. To do this, consider the following:</p>
<ul>
<li>Average Silhouette Score: Select <strong>maximum</strong> value for
best k;</li>
<li>Davies-Bouldin index: Select <strong>minimum</strong> value for best
k;<br>
</li>
<li>Calinsky-Harabasz index: Select <strong>maximum</strong> value for
best k.</li>
</ul>
<p>The most straight forward way to evaluate a reasonable range of k’s
is by using the function <code><a href="../reference/suggest_k.html">suggest_k()</a></code>. This function
includes an argument for detailed results, which will give you the
results for the three indices, between k = 3 and k = 10. We don’t
recommend testing k’s outside of this range for the purpose of defining
the rare biosphere, but you can analyze all possible k’s (just change
the range argument).</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/suggest_k.html">suggest_k</a></span><span class="op">(</span><span class="va">nice_tidy</span>, detailed <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>  </span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt; [1] "This list contains several details that might help you decide a k parameter."</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]</span></span>
<span><span class="co">#&gt;                      Score                 Criteria                     Details</span></span>
<span><span class="co">#&gt; 1     Davies-Bouldin index Minimum value for best k Measures cluster separation</span></span>
<span><span class="co">#&gt; 2  Calinsky-Harabasz index Maximum value for best k Measures cluster definition</span></span>
<span><span class="co">#&gt; 3 Average Silhouette Score Maximum value for best k    Measures cluster density</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $SamplesSummary</span></span>
<span><span class="co">#&gt; [1] "You study has 9 samples. For each one we calculated all indices obtained for each k, from 3 to 10"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $DaviesBouldin</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 9 × 3</span></span></span>
<span><span class="co">#&gt;   Sample          CH     k</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> ERR2044662  <span style="text-decoration: underline;">30</span>332.    10</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> ERR2044663  <span style="text-decoration: underline;">42</span>150.    10</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> ERR2044664 <span style="text-decoration: underline;">296</span>091.    10</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> ERR2044665  <span style="text-decoration: underline;">31</span>908.    10</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> ERR2044666  <span style="text-decoration: underline;">88</span>804.    10</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> ERR2044667  <span style="text-decoration: underline;">17</span>792.    10</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">7</span> ERR2044669  <span style="text-decoration: underline;">35</span>405.    10</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">8</span> ERR2044668  <span style="text-decoration: underline;">75</span>615.    10</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">9</span> ERR2044670  <span style="text-decoration: underline;">18</span>276.     9</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $CalinskyHarabasz</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 9 × 3</span></span></span>
<span><span class="co">#&gt;   Sample          CH     k</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> ERR2044662  <span style="text-decoration: underline;">30</span>332.    10</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> ERR2044663  <span style="text-decoration: underline;">42</span>150.    10</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> ERR2044664 <span style="text-decoration: underline;">296</span>091.    10</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> ERR2044665  <span style="text-decoration: underline;">31</span>908.    10</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> ERR2044666  <span style="text-decoration: underline;">88</span>804.    10</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> ERR2044667  <span style="text-decoration: underline;">17</span>792.    10</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">7</span> ERR2044669  <span style="text-decoration: underline;">35</span>405.    10</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">8</span> ERR2044668  <span style="text-decoration: underline;">75</span>615.    10</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">9</span> ERR2044670  <span style="text-decoration: underline;">18</span>276.     9</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $averageSilhouette</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 9 × 3</span></span></span>
<span><span class="co">#&gt;   Sample     average_Silhouette     k</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> ERR2044662              0.952     3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> ERR2044663              0.955     3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> ERR2044664              0.970     3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> ERR2044665              0.915     3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> ERR2044666              0.965     3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> ERR2044667              0.874     3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">7</span> ERR2044669              0.913     5</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">8</span> ERR2044668              0.951     4</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">9</span> ERR2044670              0.932     3</span></span></code></pre></div>
<p>For another group of k’s, range = 10:20</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/suggest_k.html">suggest_k</a></span><span class="op">(</span><span class="va">nice_tidy</span>, detailed <span class="op">=</span> <span class="cn">TRUE</span>, range <span class="op">=</span> <span class="fl">10</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span>  </span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt; [1] "This list contains several details that might help you decide a k parameter."</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]</span></span>
<span><span class="co">#&gt;                      Score                 Criteria                     Details</span></span>
<span><span class="co">#&gt; 1     Davies-Bouldin index Minimum value for best k Measures cluster separation</span></span>
<span><span class="co">#&gt; 2  Calinsky-Harabasz index Maximum value for best k Measures cluster definition</span></span>
<span><span class="co">#&gt; 3 Average Silhouette Score Maximum value for best k    Measures cluster density</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $SamplesSummary</span></span>
<span><span class="co">#&gt; [1] "You study has 9 samples. For each one we calculated all indices obtained for each k, from 10 to 20"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $DaviesBouldin</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 9 × 3</span></span></span>
<span><span class="co">#&gt;   Sample           CH     k</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> ERR2044662  <span style="text-decoration: underline;">234</span>158.    20</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> ERR2044663  <span style="text-decoration: underline;">302</span>032.    20</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> ERR2044664 2<span style="text-decoration: underline;">004</span>343.    20</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> ERR2044665  <span style="text-decoration: underline;">230</span>118.    20</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> ERR2044666  <span style="text-decoration: underline;">316</span>479.    20</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> ERR2044667  <span style="text-decoration: underline;">190</span>069.    20</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">7</span> ERR2044669  <span style="text-decoration: underline;">183</span>647.    19</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">8</span> ERR2044668  <span style="text-decoration: underline;">418</span>496.    18</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">9</span> ERR2044670  <span style="text-decoration: underline;">212</span>691.    19</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $CalinskyHarabasz</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 9 × 3</span></span></span>
<span><span class="co">#&gt;   Sample           CH     k</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> ERR2044662  <span style="text-decoration: underline;">234</span>158.    20</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> ERR2044663  <span style="text-decoration: underline;">302</span>032.    20</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> ERR2044664 2<span style="text-decoration: underline;">004</span>343.    20</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> ERR2044665  <span style="text-decoration: underline;">230</span>118.    20</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> ERR2044666  <span style="text-decoration: underline;">316</span>479.    20</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> ERR2044667  <span style="text-decoration: underline;">190</span>069.    20</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">7</span> ERR2044669  <span style="text-decoration: underline;">183</span>647.    19</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">8</span> ERR2044668  <span style="text-decoration: underline;">418</span>496.    18</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">9</span> ERR2044670  <span style="text-decoration: underline;">212</span>691.    19</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $averageSilhouette</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 9 × 3</span></span></span>
<span><span class="co">#&gt;   Sample     average_Silhouette     k</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> ERR2044662              0.770    10</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> ERR2044663              0.800    12</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> ERR2044664              0.773    10</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> ERR2044665              0.789    11</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> ERR2044666              0.781    10</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> ERR2044667              0.791    11</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">7</span> ERR2044669              0.748    12</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">8</span> ERR2044668              0.799    13</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">9</span> ERR2044670              0.747    14</span></span></code></pre></div>
<div class="section level4">
<h4 id="fine-grained-analysis">Fine grained analysis<a class="anchor" aria-label="anchor" href="#fine-grained-analysis"></a>
</h4>
<p>You can look at particular metrics and at specific samples. To do so,
we made several helper functions:</p>
<ul>
<li>
<code><a href="../reference/check_avgSil.html">check_avgSil()</a></code> to calculate average silhouette
score;</li>
<li>
<code><a href="../reference/check_DB.html">check_DB()</a></code> to calculate Davies-Bouldin index;</li>
<li>
<code><a href="../reference/check_CH.html">check_CH()</a></code>to calculate Calinsky-Harabasz index.</li>
</ul>
<p>You can also calculate them at the same time for one sample with
<code><a href="../reference/evaluate_sample_k.html">evaluate_sample_k()</a></code> or all samples with
<code><a href="../reference/evaluate_k.html">evaluate_k()</a></code>.</p>
<p>Let’s look at several examples, using the default range of k = 3 up
to k = 10:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## One sample</span></span>
<span><span class="co"># To get values</span></span>
<span><span class="fu"><a href="../reference/check_avgSil.html">check_avgSil</a></span><span class="op">(</span><span class="va">nice_tidy</span>, sample_id <span class="op">=</span> <span class="va">selected_samples</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.9521452 0.8820316 0.8561774 0.8398216 0.8479872 0.7843358 0.7740169</span></span>
<span><span class="co">#&gt; [8] 0.7701163</span></span>
<span></span>
<span><span class="co"># To plot results</span></span>
<span><span class="fu"><a href="../reference/check_avgSil.html">check_avgSil</a></span><span class="op">(</span><span class="va">nice_tidy</span>, sample_id <span class="op">=</span> <span class="va">selected_samples</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, with_plot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="explore-classifications_files/figure-html/unnamed-chunk-20-1.png" width="700"></p>
<p>In the last plot, each value of k represents a clustering result.
From the above information, we would select k = 3 (just like the default
option), because it had the highest average Silhouette score.</p>
<p>We can repeat for all other indices:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Davie-Boulding index</span></span>
<span><span class="co"># To get values</span></span>
<span><span class="fu"><a href="../reference/check_DB.html">check_DB</a></span><span class="op">(</span><span class="va">nice_tidy</span>, sample_id <span class="op">=</span> <span class="va">selected_samples</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.3721866 0.5271704 0.4131651 0.4292696 0.3350836 0.3892966 0.3948026</span></span>
<span><span class="co">#&gt; [8] 0.3294451</span></span>
<span></span>
<span><span class="co"># To plot results</span></span>
<span><span class="fu"><a href="../reference/check_DB.html">check_DB</a></span><span class="op">(</span><span class="va">nice_tidy</span>, sample_id <span class="op">=</span> <span class="va">selected_samples</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, with_plot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="explore-classifications_files/figure-html/unnamed-chunk-21-1.png" width="700"></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">## Calinsky-Harabasz index</span></span>
<span><span class="co"># To get values</span></span>
<span><span class="fu"><a href="../reference/check_CH.html">check_CH</a></span><span class="op">(</span><span class="va">nice_tidy</span>, sample_id <span class="op">=</span> <span class="va">selected_samples</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1]  1821.426  2054.887  4933.956  5465.134 17589.032 17179.809 18083.313</span></span>
<span><span class="co">#&gt; [8] 30332.345</span></span>
<span></span>
<span><span class="co"># To plot results</span></span>
<span><span class="fu"><a href="../reference/check_CH.html">check_CH</a></span><span class="op">(</span><span class="va">nice_tidy</span>, sample_id <span class="op">=</span> <span class="va">selected_samples</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, with_plot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="explore-classifications_files/figure-html/unnamed-chunk-21-2.png" width="700"></p>
<p>Note that the best k changed depending on the metric selected.</p>
<p>Thus, for sample 1 (ERR2044662), the best k would be: - 3 based on
average Silhouette scores; - 4 based on Davies-Bouldin index; - 10 based
on Calinsky-Harabasz index.</p>
<p>We can see how they compare directly with
<code><a href="../reference/evaluate_sample_k.html">evaluate_sample_k()</a></code>:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/evaluate_sample_k.html">evaluate_sample_k</a></span><span class="op">(</span><span class="va">nice_tidy</span>, sample_id <span class="op">=</span> <span class="va">selected_samples</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, with_plot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="explore-classifications_files/figure-html/unnamed-chunk-22-1.png" width="700"></p>
<p>The ulrb method assumes that samples are independent from each other,
which means that the above results for one sample, will be different for
the other samples. Thus, we should look at all samples at the same time,
with centrality metrics. To do so, we can use the function
<code><a href="../reference/evaluate_k.html">evaluate_k()</a></code>:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## To get values</span></span>
<span><span class="fu"><a href="../reference/evaluate_k.html">evaluate_k</a></span><span class="op">(</span><span class="va">nice_tidy</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 72 × 6</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Groups:   Sample [9]</span></span></span>
<span><span class="co">#&gt;    Sample     data                   DB     CH average_Silhouette     k</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> ERR2044662 <span style="color: #949494;">&lt;tibble [187 × 10]&gt;</span> 0.372  <span style="text-decoration: underline;">1</span>821.              0.952     3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> ERR2044662 <span style="color: #949494;">&lt;tibble [187 × 10]&gt;</span> 0.527  <span style="text-decoration: underline;">2</span>055.              0.882     4</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> ERR2044662 <span style="color: #949494;">&lt;tibble [187 × 10]&gt;</span> 0.413  <span style="text-decoration: underline;">4</span>934.              0.856     5</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> ERR2044662 <span style="color: #949494;">&lt;tibble [187 × 10]&gt;</span> 0.429  <span style="text-decoration: underline;">5</span>465.              0.840     6</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> ERR2044662 <span style="color: #949494;">&lt;tibble [187 × 10]&gt;</span> 0.335 <span style="text-decoration: underline;">17</span>589.              0.848     7</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> ERR2044662 <span style="color: #949494;">&lt;tibble [187 × 10]&gt;</span> 0.389 <span style="text-decoration: underline;">17</span>180.              0.784     8</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> ERR2044662 <span style="color: #949494;">&lt;tibble [187 × 10]&gt;</span> 0.395 <span style="text-decoration: underline;">18</span>083.              0.774     9</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> ERR2044662 <span style="color: #949494;">&lt;tibble [187 × 10]&gt;</span> 0.329 <span style="text-decoration: underline;">30</span>332.              0.770    10</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> ERR2044663 <span style="color: #949494;">&lt;tibble [220 × 10]&gt;</span> 0.391  <span style="text-decoration: underline;">1</span>491.              0.955     3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> ERR2044663 <span style="color: #949494;">&lt;tibble [220 × 10]&gt;</span> 0.516  <span style="text-decoration: underline;">1</span>495.              0.891     4</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 62 more rows</span></span></span>
<span></span>
<span><span class="co">## To plot</span></span>
<span><span class="fu"><a href="../reference/evaluate_k.html">evaluate_k</a></span><span class="op">(</span><span class="va">nice_tidy</span>, with_plot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; No summary function supplied, defaulting to `mean_se()`</span></span>
<span><span class="co">#&gt; No summary function supplied, defaulting to `mean_se()`</span></span>
<span><span class="co">#&gt; No summary function supplied, defaulting to `mean_se()`</span></span></code></pre></div>
<p><img src="explore-classifications_files/figure-html/unnamed-chunk-23-1.png" width="700"></p>
</div>
</div>
<div class="section level3">
<h3 id="automatic-k-selection">Automatic k selection<a class="anchor" aria-label="anchor" href="#automatic-k-selection"></a>
</h3>
<p>We can decide what k to use based on any of this metrics, being aware
that they measure different aspects of the clustering results. The
function <code><a href="../reference/suggest_k.html">suggest_k()</a></code>, by default, will:</p>
<ul>
<li>calculate the best k, based on the average Silhouette score, for
<strong>each sample</strong>;</li>
<li>calculate the average best k across all samples;</li>
<li>return best k as an integer.</li>
</ul>
<p>Thus, the default output of <code><a href="../reference/suggest_k.html">suggest_k()</a></code> is a single
integer; this is used in the <code><a href="../reference/define_rb.html">define_rb()</a></code> function for the
automatic k decision.</p>
<p>Instead of the average Silhouette score, you can select another index
(Davies-Bouldin or Calinsky-Harabasz), and the function
<code><a href="../reference/suggest_k.html">suggest_k()</a></code> will return the best k based on that index.</p>
<p>Let’s see some examples:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># default option with average Silhouette score</span></span>
<span><span class="fu"><a href="../reference/suggest_k.html">suggest_k</a></span><span class="op">(</span><span class="va">nice_tidy</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 3</span></span>
<span></span>
<span><span class="co"># best k for Davies-Bouldin</span></span>
<span><span class="fu"><a href="../reference/suggest_k.html">suggest_k</a></span><span class="op">(</span><span class="va">nice_tidy</span>, index <span class="op">=</span> <span class="st">"Davies-Bouldin"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 6</span></span>
<span></span>
<span><span class="co"># best k for Calinsky-Harabasz</span></span>
<span><span class="fu"><a href="../reference/suggest_k.html">suggest_k</a></span><span class="op">(</span><span class="va">nice_tidy</span>, index <span class="op">=</span> <span class="st">"Calinsky-Harabasz"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 9</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="everything-automatic">Everything automatic<a class="anchor" aria-label="anchor" href="#everything-automatic"></a>
</h3>
<p>Finally, if we have the ability to automatically suggest the value of
k, we can do the same in the context of the definition of rarity. To so
do, you can use the default parameters in <code><a href="../reference/define_rb.html">define_rb()</a></code>, but
with the automatic argument set to TRUE.</p>
<p>Like so,</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">automatic_classification</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/define_rb.html">define_rb</a></span><span class="op">(</span><span class="va">nice_tidy</span>, automatic <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Automatic option set to TRUE, so classification vector was overwritten</span></span>
<span><span class="co">#&gt; K= 3 based on Average Silhouette Score.</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(Sample, Level)`</span></span>
<span></span>
<span><span class="co"># Plot automatic result</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot_ulrb.html">plot_ulrb</a></span><span class="op">(</span><span class="va">automatic_classification</span>, taxa_col <span class="op">=</span> <span class="st">"OTU"</span>, plot_all <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 197 rows containing missing values (`geom_segment()`).</span></span>
<span><span class="co">#&gt; Removed 197 rows containing missing values (`geom_segment()`).</span></span></code></pre></div>
<p><img src="explore-classifications_files/figure-html/unnamed-chunk-25-1.png" width="700"></p>
<p>Naturally, you can decide more parameters for the automatic selection
of k, for example, lets suppose that you want to have groups of
phylogenetic units that are very well defined, but you also want to
have, at least, four classifications, but no more than 6.</p>
<p>For those two conditions, you would have to specify the range of k
values (4:6) and the evaluation index should be the Calinsky-Harabasz
index:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">more_complex_automatic_classification</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/define_rb.html">define_rb</a></span><span class="op">(</span><span class="va">nice_tidy</span>, </span>
<span>                                                   automatic <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                                   index <span class="op">=</span> <span class="st">"Calinsky-Harabasz"</span>,</span>
<span>                                                   range <span class="op">=</span> <span class="fl">4</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span></span>
<span><span class="co">#&gt; Automatic option set to TRUE, so classification vector was overwritten</span></span>
<span><span class="co">#&gt; K= 5 based on Calinsky-Harabasz.</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(Sample, Level)`</span></span></code></pre></div>
<p>Note that the function informed you that the automatic k selected was
5. Thus, you know you will need 5 colors for the standard ulrb
plots:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot automatic result</span></span>
<span><span class="fu"><a href="../reference/plot_ulrb.html">plot_ulrb</a></span><span class="op">(</span><span class="va">more_complex_automatic_classification</span>, </span>
<span>          plot_all <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>          taxa_col <span class="op">=</span> <span class="st">"OTU"</span>, </span>
<span>          colors <span class="op">=</span> <span class="va">qualitative_colors</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>,</span>
<span>          log_scaled <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 213 rows containing missing values (`geom_segment()`).</span></span>
<span><span class="co">#&gt; Removed 213 rows containing missing values (`geom_segment()`).</span></span></code></pre></div>
<p><img src="explore-classifications_files/figure-html/unnamed-chunk-27-1.png" width="700"></p>
<p>In this example, the automatic option did not seem to give a better
result than the default k = 3.</p>
</div>
<div class="section level3">
<h3 id="how-each-index-behaves-across-all-possible-values-of-k">How each index behaves across all possible values of k?<a class="anchor" aria-label="anchor" href="#how-each-index-behaves-across-all-possible-values-of-k"></a>
</h3>
<p>This last section is for completion sake and should not be necessary
in a study of the rare biosphere; however, it might prove useful for
those interested in the details of this unsupervised learning
approach.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Start by deciding the maximum range across the entire dataset</span></span>
<span><span class="va">max_k</span> <span class="op">&lt;-</span> <span class="va">nice_tidy</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">Abundance</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">Abundance</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">Sample</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>topK <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">Abundance</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">ungroup</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">topK</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># print maximum number of clusters allowed for all samples in the N-ICE dataset</span></span>
<span><span class="va">max_k</span></span>
<span><span class="co">#&gt; [1] 55</span></span></code></pre></div>
<p>If we have the maximum k, we just need to use the
<code><a href="../reference/evaluate_k.html">evaluate_k()</a></code> function for the full range of values. (if you
try to run this code, it might take a while)</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/evaluate_k.html">evaluate_k</a></span><span class="op">(</span><span class="va">nice_tidy</span>, with_plot <span class="op">=</span> <span class="cn">TRUE</span>, range <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="va">max_k</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 1 rows containing non-finite values (`stat_summary()`).</span></span>
<span><span class="co">#&gt; No summary function supplied, defaulting to `mean_se()`</span></span>
<span><span class="co">#&gt; No summary function supplied, defaulting to `mean_se()`</span></span>
<span><span class="co">#&gt; No summary function supplied, defaulting to `mean_se()`</span></span></code></pre></div>
<p><img src="explore-classifications_files/figure-html/unnamed-chunk-29-1.png" width="700"></p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Francisco Pascoal, Paula Branco, Luís Torgo, Rodrigo Costa, Catarina Magalhães.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
