<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="ulrb">
<title>Tutorial to define rare biosphere with ulrb • ulrb</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Tutorial to define rare biosphere with ulrb">
<meta property="og:description" content="ulrb">
<meta property="og:image" content="https://pascoalf.github.io/ulrb/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">ulrb</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/Glossary.html">Glossary</a>
    <a class="dropdown-item" href="../articles/eco-analysis.html">Integration of ulrb in a simple microbial ecology workflow</a>
    <a class="dropdown-item" href="../articles/explore-classifications.html">Alternative classifications with ulrb</a>
    <a class="dropdown-item" href="../articles/ulrb-vignet.html">Tutorial to define rare biosphere with ulrb</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/pascoalf/ulrb/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Tutorial to define rare biosphere with ulrb</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/pascoalf/ulrb/blob/HEAD/vignettes/ulrb-vignet.Rmd" class="external-link"><code>vignettes/ulrb-vignet.Rmd</code></a></small>
      <div class="d-none name"><code>ulrb-vignet.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="unsupervised-learning-based-definition-of-microbial-rare-biosphere">Unsupervised Learning Based Definition Of Microbial Rare
Biosphere<a class="anchor" aria-label="anchor" href="#unsupervised-learning-based-definition-of-microbial-rare-biosphere"></a>
</h2>
<p>The R package ulrb solves the problem of the definition of rarity by
replacing human decision with an unsupervised machine learning algorithm
(partitioning around medoids, k-medoids). This algorithm works for any
type of microbiome data, provided there is an abundance score for each
phylogenetic unit. For validation of this method to several kinds of
molecular data and environments, please see Pascoal et al., 2023 (in
preparation). Preliminary results indicate that this method works for
oter biological systems, provided there is an abundance table.</p>
<p>In this tutorial we will illustrate how to define the microbial rare
biosphere with a small, publicly available dataset, the Norwegian Young
Sea Ice Expedition, 2016 (N-ICE). The molecular data will be from
amplicon sequencing of 16S rRNA gene (for sequencing details, see Sousa
et al., 2019), raw sequences are available in the European Nucleotide
Archive (ENA) under accession number ERP024265, and they were processed
into OTUs by MGnify (Mitchell et al.,2019).</p>
<p>The N-ICE dataset consists of 9 samples, collected during the winter
to spring transition of 2015 by fixing the vessel on drifting ice
(Granskog et al., 2018). Samples were collected at various depths (from
5m to 250m) and the ice drifted across different regions, DNA was
collected for 16S amplicon sequencing and metagenomes (de Sousa et al.,
2019); and bioinformatic processing of reads was performed by Mgnify
platform (v5) (Mitchel et al., 2020).</p>
<p>For this tutorial we are using the OTU table downloaded from the link
<a href="https://www.ebi.ac.uk/metagenomics/studies/MGYS00001922#analysis" class="external-link uri">https://www.ebi.ac.uk/metagenomics/studies/MGYS00001922#analysis</a>
in 06-01-2023 and focus solely on the 16S amplicon data.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://pascoalf.github.io/ulrb/">ulrb</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org" class="external-link">tidyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="brief-note-on-nomenclature">Brief note on nomenclature<a class="anchor" aria-label="anchor" href="#brief-note-on-nomenclature"></a>
</h4>
<p>We use the term “phylogenetic units” to refer to the highest
resolution, distinct unit identified. This can be species, OTU, ASV,
…</p>
<p>We use the term “abundance table” to refer to the table with, at
least, the abundance score for each phylogenetic unit. Any additional
data can be added, this can be an OTU table, ASV table…</p>
<p>We also use the term “scale” or “scaling”, which can also appear in
the literature as “normalization”, “transformation”, …</p>
</div>
<div class="section level3">
<h3 id="pre-processing-of-data-prior-to-clustering-algorithm">Pre-processing of data prior to clustering algorithm<a class="anchor" aria-label="anchor" href="#pre-processing-of-data-prior-to-clustering-algorithm"></a>
</h3>
<p>The package ulrb will require that the abundance table is in tidy (or
long) format, with a column for the abundance score used and another
column for samples. Any additional columns are allowed.</p>
<p><strong>Regarding the abundance score</strong>, we are not scaling
the abundance score in this vignette, because the k-medoids algorithm
will be applied “per sample” to the values of the abundance score. This
means that any scaling, <em>e.g.</em> relative abundance, center
log-ratio, … Will not have an effect on the algorithm, because the
distance values between observations will remain the same. However, if
the algorithm was applied to more variables (not only abundance), then
scaling could make a difference. Thus, we consider that it is best to
use absolute values in this tutorial, to avoid misleading the reader
into thinking that some scaling has to be done.</p>
<p>Notwithstanding, for publication purposes and for the purposes of
other data processing steps, it might make sense to apply some sort of
scaling. As long as the number of observations, <em>i.e.</em> the number
of phylogenetic units, is the same, then the distance between them will
be the same.</p>
<p>If the user decides to apply any quality filtering to the abundance
table, like rarefaction, it should be done <strong>before</strong> using
the clustering method.</p>
<p>In summary, the algorithm will be applied to your final, clean
abundance table, with any abundance score.</p>
<div class="section level4">
<h4 id="load-and-clean-abundance-table">Load and clean abundance table<a class="anchor" aria-label="anchor" href="#load-and-clean-abundance-table"></a>
</h4>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load raw OTU table from N-ICE</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="st">"../data/nice_raw.rda"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Change name of first column</span></span>
<span><span class="va">nice_clean</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span><span class="va">nice_raw</span>, Taxonomy <span class="op">=</span> <span class="st">"X.SampleID"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Select 16S rRNA amplicon sequencing samples</span></span>
<span><span class="va">selected_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ERR2044662"</span>, <span class="st">"ERR2044663"</span>, <span class="st">"ERR2044664"</span>,<span class="st">"ERR2044665"</span>, <span class="st">"ERR2044666"</span>, <span class="st">"ERR2044667"</span>,<span class="st">"ERR2044668"</span>, <span class="st">"ERR2044669"</span>, <span class="st">"ERR2044670"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add a column with phylogenetic units ID (OTU in this case)</span></span>
<span><span class="va">nice_clean</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span><span class="va">nice_clean</span>, OTU <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"OTU_"</span>, <span class="fu"><a href="https://dplyr.tidyverse.org/reference/row_number.html" class="external-link">row_number</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Select relevant columns</span></span>
<span><span class="va">nice_clean</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">nice_clean</span>, <span class="va">selected_samples</span>, <span class="va">OTU</span>, <span class="va">Taxonomy</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Using an external vector in selections was deprecated in tidyselect 1.1.0.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Please use `all_of()` or `any_of()` instead.</span></span>
<span><span class="co">#&gt;   # Was:</span></span>
<span><span class="co">#&gt;   data %&gt;% select(selected_samples)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   # Now:</span></span>
<span><span class="co">#&gt;   data %&gt;% select(all_of(selected_samples))</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; See &lt;https://tidyselect.r-lib.org/reference/faq-external-vector.html&gt;.</span></span>
<span><span class="co">#&gt; <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">generated.</span></span></span>
<span></span>
<span><span class="co"># Separate Taxonomy column into each taxonomic level</span></span>
<span><span class="va">nice_clean</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tidyr.tidyverse.org/reference/separate.html" class="external-link">separate</a></span><span class="op">(</span><span class="va">nice_clean</span>, <span class="va">Taxonomy</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Domain"</span>,<span class="st">"Kingdom"</span>,<span class="st">"Phylum"</span>,<span class="st">"Class"</span>,<span class="st">"Order"</span>,<span class="st">"Family"</span>,<span class="st">"Genus"</span>,<span class="st">"Species"</span><span class="op">)</span>,sep<span class="op">=</span><span class="st">";"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Expected 8 pieces. Missing pieces filled with `NA` in 912 rows [1, 2, 4, 5, 6,</span></span>
<span><span class="co">#&gt; 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, ...].</span></span>
<span></span>
<span><span class="co"># Remove Kingdom column, because it is not used for prokaryotes</span></span>
<span><span class="va">nice_clean</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">nice_clean</span>, <span class="op">-</span><span class="va">Kingdom</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Remove eukaryotes</span></span>
<span><span class="va">nice_clean</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">nice_clean</span>, <span class="va">Domain</span> <span class="op">!=</span> <span class="st">"sk__Eukaryota"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Remove unclassified OTUs at phylum level</span></span>
<span><span class="va">nice_clean</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">nice_clean</span>, <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">Phylum</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Simplify name</span></span>
<span><span class="va">nice</span> <span class="op">&lt;-</span> <span class="va">nice_clean</span></span>
<span></span>
<span><span class="co"># Quick look at the table</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">nice</span><span class="op">)</span></span>
<span><span class="co">#&gt;   ERR2044662 ERR2044663 ERR2044664 ERR2044665 ERR2044666 ERR2044667 ERR2044668</span></span>
<span><span class="co">#&gt; 1        165        323         51         70        134        216          0</span></span>
<span><span class="co">#&gt; 2          0          0          1          0          0          1          0</span></span>
<span><span class="co">#&gt; 3          0          0          1          2          2          6          0</span></span>
<span><span class="co">#&gt; 4        541       1018        351        115        241       1633        177</span></span>
<span><span class="co">#&gt; 5          8          5         41         15         14        146          0</span></span>
<span><span class="co">#&gt; 6         15         31        590        133        174       1814         12</span></span>
<span><span class="co">#&gt;   ERR2044669 ERR2044670   OTU      Domain            Phylum</span></span>
<span><span class="co">#&gt; 1         11          0 OTU_2 sk__Archaea  p__Euryarchaeota</span></span>
<span><span class="co">#&gt; 2          0          0 OTU_3 sk__Archaea  p__Euryarchaeota</span></span>
<span><span class="co">#&gt; 3          0          0 OTU_4 sk__Archaea  p__Euryarchaeota</span></span>
<span><span class="co">#&gt; 4       1371          7 OTU_5 sk__Archaea  p__Euryarchaeota</span></span>
<span><span class="co">#&gt; 5         14          0 OTU_6 sk__Archaea p__Thaumarchaeota</span></span>
<span><span class="co">#&gt; 6        173          2 OTU_7 sk__Archaea p__Thaumarchaeota</span></span>
<span><span class="co">#&gt;                       Class                       Order Family</span></span>
<span><span class="co">#&gt; 1 c__Candidatus_Poseidoniia                        &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 2 c__Candidatus_Poseidoniia o__Candidatus_Poseidoniales    f__</span></span>
<span><span class="co">#&gt; 3           c__Halobacteria          o__Halobacteriales   &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 4         c__Thermoplasmata                        &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 5                      &lt;NA&gt;                        &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 6                       c__                         o__    f__</span></span>
<span><span class="co">#&gt;                            Genus                                        Species</span></span>
<span><span class="co">#&gt; 1                           &lt;NA&gt;                                           &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 2                            g__ s__Marine_group_II_euryarchaeote_REDSEA-S03_B6</span></span>
<span><span class="co">#&gt; 3                           &lt;NA&gt;                                           &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 4                           &lt;NA&gt;                                           &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 5                           &lt;NA&gt;                                           &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 6 g__Candidatus_Nitrosopelagicus                                           &lt;NA&gt;</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="transform-abundance-table-into-tidylong-format">Transform abundance table into tidy/long format<a class="anchor" aria-label="anchor" href="#transform-abundance-table-into-tidylong-format"></a>
</h4>
<p>Now that we have a cleaned OTU table, we can transform it to tidy
format. You can do this part in any order and according to your own data
format. We made the function <code><a href="../reference/prepare_tidy_data.html">prepare_tidy_data()</a></code> to help
transform the two most common data formats for abundance tables:</p>
<ol style="list-style-type: decimal">
<li>phylogenetic units in rows, samples in columns;</li>
<li>phylogenetic units in columns, samples in rows.</li>
</ol>
<p>In this example, we have a data.frame with phylogenetic units as rows
and samples in various columns, so this is not a tidy format. We need to
collapse all columns with samples into a single “Abundance” and
“Samples” column, so that each variable corresponds to a single column,
<em>i.e.</em> tidy.</p>
<p>For this we just need to apply the <code><a href="../reference/prepare_tidy_data.html">prepare_tidy_data()</a></code>
function to the cleaned data. Note that you need to specify in which
format your abundance table is. In this example, samples are in
different columns, so we set the argument samples_in to “cols”; if it
was the opposite situation, with samples in rows, we would set
samples_in to “rows”.</p>
<p>This function is optional and if you use it you really should verify
if it is transforming the data correctly, because your own data might be
different from our standard examples.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nice_tidy</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepare_tidy_data.html">prepare_tidy_data</a></span><span class="op">(</span><span class="va">nice</span>, sample_names <span class="op">=</span> <span class="va">selected_samples</span>, samples_in <span class="op">=</span> <span class="st">"cols"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="apply-definition-of-rare-biosphere-with-unsupervised-learning">Apply definition of rare biosphere with unsupervised learning<a class="anchor" aria-label="anchor" href="#apply-definition-of-rare-biosphere-with-unsupervised-learning"></a>
</h3>
<p>Now we are ready to use the central function of the package ulrb,
which is <code><a href="../reference/define_rb.html">define_rb()</a></code>.</p>
<p>This function will add new columns to the input data.frame,
automatically deciding the classification of each phylogenetic unit, per
sample, by means of the partition around medoids algorithm.</p>
<p>The algorithm is run by the <code><a href="https://rdrr.io/pkg/cluster/man/pam.html" class="external-link">pam()</a></code> function of the
<code>cluster</code> package. The output are the clusters, which we
translate into classifications, and provides statistical scores to
interpret the robustness of the clusters formed. From those statistics,
we select the silhouette scores, which will be used to evaluate the
results later on. Additionally, the data.frame will include a nested
list with the output from <code><a href="https://rdrr.io/pkg/cluster/man/pam.html" class="external-link">pam()</a></code>, for more experienced users
to analyze if necessary.</p>
<p>Regarding the classifications, we are setting a default to “rare”,
“undetermined” and “abundant”, based on Pascoal et al., 2023 (manuscript
in preparation). However, the user can set any vector of
classifications, as long as they follow the rules of the number of
clusters, or k, from <code><a href="https://rdrr.io/pkg/cluster/man/pam.html" class="external-link">pam()</a></code>. Meaning, the number of
classifications can’t be more than the number of distinct observations
in the data. In practice this is not a concern, because we think most
researchers will either use our default, or “rare” vs “abundant”, or a
more nuanced “very rare”, “rare”, “abundant”, “very abundant”. The point
is that the classification can be decided in a study by study
context.</p>
<p>Since the default output includes so many columns that are there to
support the classification obtained, we added an argument, simplified,
which if TRUE will return a simplified version of the output, without
silhouette scores, just the final classification, plus all the input
data. We will use the default, simplified = FALSE.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">classified_table</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/define_rb.html">define_rb</a></span><span class="op">(</span><span class="va">nice_tidy</span><span class="op">)</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(Sample, Level)`</span></span>
<span></span>
<span><span class="co"># Quick output check</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">classified_table</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "Sample"                   "Classification"          </span></span>
<span><span class="co">#&gt;  [3] "OTU"                      "Domain"                  </span></span>
<span><span class="co">#&gt;  [5] "Phylum"                   "Class"                   </span></span>
<span><span class="co">#&gt;  [7] "Order"                    "Family"                  </span></span>
<span><span class="co">#&gt;  [9] "Genus"                    "Species"                 </span></span>
<span><span class="co">#&gt; [11] "Abundance"                "pam_object"              </span></span>
<span><span class="co">#&gt; [13] "Level"                    "Silhouette_scores"       </span></span>
<span><span class="co">#&gt; [15] "Cluster_median_abundance" "median_Silhouette"       </span></span>
<span><span class="co">#&gt; [17] "Evaluation"</span></span>
<span></span>
<span><span class="va">classified_table</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">OTU</span>, <span class="va">Sample</span>, <span class="va">Abundance</span>, </span>
<span>         <span class="va">Classification</span>, <span class="va">Silhouette_scores</span>, <span class="va">Cluster_median_abundance</span>, </span>
<span>         <span class="va">pam_object</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 7</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Groups:   Sample, Classification [1]</span></span></span>
<span><span class="co">#&gt;   OTU   Sample Abundance Classification Silhouette_scores Cluster_median_abund…¹</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>                      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> OTU_2 ERR20…       165 Rare                       0.985                    5.5</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> OTU_5 ERR20…       541 Rare                       0.985                    5.5</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> OTU_6 ERR20…         8 Rare                       0.985                    5.5</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> OTU_7 ERR20…        15 Rare                       0.985                    5.5</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> OTU_8 ERR20…         5 Rare                       0.985                    5.5</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> OTU_… ERR20…         4 Rare                       0.985                    5.5</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ abbreviated name: ¹​Cluster_median_abundance</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1 more variable: pam_object &lt;list&gt;</span></span></span></code></pre></div>
<div class="section level4">
<h4 id="fully-automated-version">Fully automated version<a class="anchor" aria-label="anchor" href="#fully-automated-version"></a>
</h4>
<p>In the previous example, there is still one level of subjective
decision, which is the number of classifications to use (<em>e.g.</em>
“rare”, “undetermined” and “abundant” would be k = 3). However, we added
an option to the main function, <code><a href="../reference/define_rb.html">define_rb()</a></code>, which allows
for the automatic decision of the number of clusters. This automation
works under the assumption that the range of good k values lies between
3 and 10 (see Pascoal et al., 2023 for reasoning behind this). In
<code><a href="../articles/explore-classifications.html">vignette("explore-classifications")</a></code>, we explain in more
detail how the automation works and how you can make your own decision
regarding k, using several metrics.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Simple automation example</span></span>
<span><span class="fu"><a href="../reference/define_rb.html">define_rb</a></span><span class="op">(</span><span class="va">nice_tidy</span>, automatic <span class="op">=</span>  <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Automatic option set to TRUE, so classification vector was overwritten</span></span>
<span><span class="co">#&gt; K= 3 based on Average Silhouette Score.</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(Sample, Level)`</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 2,177 × 17</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Groups:   Sample, Classification [27]</span></span></span>
<span><span class="co">#&gt;    Sample    Classification OTU   Domain Phylum Class Order Family Genus Species</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> ERR20446… 1              OTU_2 sk__A… p__Eu… c__C… <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> ERR20446… 1              OTU_5 sk__A… p__Eu… c__T… <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> ERR20446… 1              OTU_6 sk__A… p__Th… <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> ERR20446… 1              OTU_7 sk__A… p__Th… c__   o__   f__    g__C… <span style="color: #BB0000;">NA</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> ERR20446… 1              OTU_8 sk__A… p__Th… c__   o__N… <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> ERR20446… 1              OTU_… sk__A… p__Th… c__   o__N… f__Ni… g__N… <span style="color: #BB0000;">NA</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> ERR20446… 1              OTU_… sk__B… p__Ac… <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> ERR20446… 1              OTU_… sk__B… p__Ac… c__A… o__A… <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> ERR20446… 1              OTU_… sk__B… p__Ac… <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> ERR20446… 1              OTU_… sk__B… p__Ac… c__A… <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 2,167 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 7 more variables: Abundance &lt;int&gt;, pam_object &lt;list&gt;, Level &lt;fct&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   Silhouette_scores &lt;dbl&gt;, Cluster_median_abundance &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   median_Silhouette &lt;dbl&gt;, Evaluation &lt;chr&gt;</span></span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="verify-results">Verify results<a class="anchor" aria-label="anchor" href="#verify-results"></a>
</h3>
<p>It is good practice to check if results from unsupervised learning
look sane. To help with this task, we made a group of functions that
will help researchers understand if they need more or less
classifications (change the value of k), or if the algorithm just
doesn’t work. There is no hard rule to verify unsupervised learning, but
results should look reasonable and follow expectations. We have seen
that the results from this method are consistent with previous
definitions of rarity, for example, most phylogenetic units are rare and
only very few are abundant, but we avoid introducing a human decision on
which points are rare or abundant.</p>
<p>The plots are the (1) Rank Abundance Curve and the (2) Silhouette
plots. The first will be more helpful for microbial ecologists, while
the second will be more helpful for data scientists.</p>
<p>Both plots are made available by the function
<code><a href="../reference/plot_ulrb.html">plot_ulrb()</a></code> (to verify clustering and Silhouette score); or
<code><a href="../reference/plot_ulrb_clustering.html">plot_ulrb_clustering()</a></code> and
<code><a href="../reference/plot_ulrb_silhouette.html">plot_ulrb_silhouette()</a></code>, to verify the clustering or the
Silhouette, respectively.</p>
<div class="section level4">
<h4 id="rank-abunddance-curve-rac-to-verify-clustering">(1) Rank Abunddance Curve (RAC) to verify clustering<a class="anchor" aria-label="anchor" href="#rank-abunddance-curve-rac-to-verify-clustering"></a>
</h4>
<p>The RAC is a common tool in microbial ecology to describe the
detection limits of different methods and the distribution of species
abundance (Pedrós-Alió, 2012) - it ranks all species from most to least
abundant, resulting in a few species with very high abundance and a
“long-tail” of low abundance species (which are called the “rare
biosphere”). However, there is no objective method to decide where the
“long-tail” begins.</p>
<p>The ulrb method considers that the terms “rare” and “abundant” are
relative, i.e. one phylogenetic unit is rare if, and only if, compared
to an abundant one (and vice-versa). We explore the consequences and
assumptions that play into this in Pascoal et al., 2023 (manuscript in
preparation). But this means that the method is not looking for the
beginning of the “long-tail”, it is comparing the abundance of all
phylogenetic units, and clustering them accordingly.</p>
<p>Thus, if we plot the standard RAC and then color each phylogenetic
unit according with the unsupervised classification, we can check if the
unsupervised classification is consistent with the notion that the
rare-biosphere is within the “long-tail”.</p>
<p>To do this, we provide the function
<code><a href="../reference/plot_ulrb_clustering.html">plot_ulrb_clustering()</a></code>. Again, the abundance score can be
transformed and the Log10 scale can be applied optionally. Other ggplot2
functions and arguments can be used, or you can also make your own plots
and analysis.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># One sample as example</span></span>
<span><span class="fu"><a href="../reference/plot_ulrb_clustering.html">plot_ulrb_clustering</a></span><span class="op">(</span><span class="va">classified_table</span>, </span>
<span>                       sample_id <span class="op">=</span> <span class="va">selected_samples</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>                       taxa_col <span class="op">=</span> <span class="st">"OTU"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Clustering for sample"</span>, <span class="va">selected_samples</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="ulrb-vignet_files/figure-html/unnamed-chunk-6-1.png" width="576"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># All samples, with centrality metric</span></span>
<span><span class="fu"><a href="../reference/plot_ulrb_clustering.html">plot_ulrb_clustering</a></span><span class="op">(</span><span class="va">classified_table</span>,</span>
<span>                     taxa_col <span class="op">=</span> <span class="st">"OTU"</span>, </span>
<span>                     plot_all <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                     log_scaled <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Clustering for all samples"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 197 rows containing missing values (`geom_segment()`).</span></span></code></pre></div>
<p><img src="ulrb-vignet_files/figure-html/unnamed-chunk-6-2.png" width="576"></p>
<p>From the RAC obtained, we confirm that the unsupervised
classification provides a reasonable result. Because the rare biosphere
is indeed in the “long-tail”, followed by a transition region, which we
term “undetermined”, because they are neither rare, more abundant, and a
few extreme points which are abundant.</p>
</div>
<div class="section level4">
<h4 id="silhouette-plots">(2) Silhouette plots<a class="anchor" aria-label="anchor" href="#silhouette-plots"></a>
</h4>
<p>The Silhouette plot provides a notion of the coherence of the points
inside each cluster. The values for each observation range between -1
and +1, the closer to +1, the more coherent is the cluster. Usually,
most observations will be very close to +1, and a few will have low
values. As a rule of thumb, if most values are above 0.75, then the
clusters are strong, but if you have many observations close to 0, but
positive, the cluster might be artificial; for negative values, then the
clusters have no structure.</p>
<p>For more details see the documentation from the functions
<code><a href="../reference/plot_ulrb_silhouette.html">plot_ulrb_silhouette()</a></code> and <code><a href="../reference/check_avgSil.html">check_avgSil()</a></code>; we
also refer to Chapter 2 of “Finding Groups in Data: An Introduction to
Cluster Analysis” (Kaufman and Rousseuw, 1991), for more details on
interpretation of Silhouette plots.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># One sample as example</span></span>
<span><span class="fu"><a href="../reference/plot_ulrb_silhouette.html">plot_ulrb_silhouette</a></span><span class="op">(</span><span class="va">classified_table</span>,</span>
<span>                     sample_id <span class="op">=</span> <span class="va">selected_samples</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>                     taxa_col <span class="op">=</span> <span class="st">"OTU"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Silhouette plot of sample"</span>, <span class="va">selected_samples</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="ulrb-vignet_files/figure-html/unnamed-chunk-7-1.png" width="960"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">#</span></span>
<span><span class="fu"><a href="../reference/plot_ulrb_silhouette.html">plot_ulrb_silhouette</a></span><span class="op">(</span><span class="va">classified_table</span>,</span>
<span>                     sample_id <span class="op">=</span> <span class="va">selected_samples</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>                     taxa_col <span class="op">=</span> <span class="st">"OTU"</span>,</span>
<span>                     plot_all <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Silhouette plot of all samples"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 197 rows containing missing values (`geom_segment()`).</span></span></code></pre></div>
<p><img src="ulrb-vignet_files/figure-html/unnamed-chunk-7-2.png" width="960"></p>
<p>As expected, most points were very close to +1, with very few points
falling below 0.5.</p>
</div>
</div>
<div class="section level3">
<h3 id="sanity-check-summary">Sanity check summary<a class="anchor" aria-label="anchor" href="#sanity-check-summary"></a>
</h3>
<p>We provide the function <code><a href="../reference/plot_ulrb.html">plot_ulrb()</a></code> that wraps around
<code><a href="../reference/plot_ulrb_clustering.html">plot_ulrb_clustering()</a></code> and
<code><a href="../reference/plot_ulrb_silhouette.html">plot_ulrb_silhouette()</a></code> to plot the RAC and Silhouette plot
in the same grid. This function works for a single sample or for all
samples.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># For a single sample</span></span>
<span><span class="fu"><a href="../reference/plot_ulrb.html">plot_ulrb</a></span><span class="op">(</span><span class="va">classified_table</span>,</span>
<span>          sample_id <span class="op">=</span>  <span class="va">selected_samples</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>          taxa_col <span class="op">=</span> <span class="st">"OTU"</span><span class="op">)</span></span></code></pre></div>
<p><img src="ulrb-vignet_files/figure-html/unnamed-chunk-8-1.png" width="960"></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># For all samples</span></span>
<span><span class="fu"><a href="../reference/plot_ulrb.html">plot_ulrb</a></span><span class="op">(</span><span class="va">classified_table</span>,</span>
<span>          taxa_col <span class="op">=</span> <span class="st">"OTU"</span>,</span>
<span>          plot_all <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> </span>
<span><span class="co">#&gt; Warning: Removed 197 rows containing missing values (`geom_segment()`).</span></span>
<span><span class="co">#&gt; Removed 197 rows containing missing values (`geom_segment()`).</span></span></code></pre></div>
<p><img src="ulrb-vignet_files/figure-html/unnamed-chunk-8-2.png" width="960"></p>
</div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<ul>
<li><p>Pascoal et al., 2023 (manuscript in preparation)</p></li>
<li><p>Granskog, M. A., Fer, I., Rinke, A., &amp; Steen, H. (2018).
Atmosphere-Ice-Ocean-Ecosystem Processes in a Thinner Arctic Sea Ice
Regime: The Norwegian Young Sea ICE (N-ICE2015) Expedition. Journal of
Geophysical Research: Oceans, 123(3), 1586–1594.</p></li>
<li><p>de Sousa, A. G. G., Tomasino, M. P., Duarte, P.,
Fernández-Méndez, M., Assmy, P., Ribeiro, H., Surkont, J., Leite, R. B.,
Pereira-Leal, J. B., Torgo, L., Magalhães, C. (2019). Diversity and
Composition of Pelagic Prokaryotic and Protist Communities in a Thin
Arctic Sea-Ice Regime. <i>Microbial Ecology</i>, <i>78</i>(2),
388–408.</p></li>
<li><p>Mitchell, A. L., Almeida, A., Beracochea, M., Boland, M., Burgin,
J., Cochrane, G., Crusoe, M. R., Kale, V., Potter, S. C., Richardson, L.
J., Sakharova, E., Scheremetjew, M., Korobeynikov, A., Shlemov, A.,
Kunyavskaya, O., Lapidus, A., Finn, R. D. (2019). MGnify: the microbiome
analysis resource in 2020. <i>Nucleic Acids Research</i>, <i>48</i>,
D570–D578.</p></li>
<li><p>Pedrós-Alió, C. (2012). The Rare Bacterial Biosphere. <i>Annual
Review of Marine Science</i>, <i>4</i>(1), 449–466.</p></li>
<li><p>Kaufman, L., &amp; Rousseuw, P. J. (1991). Chapter 2 in book
Finding Groups in Data: An Introduction to Cluster Analysis. Biometrics,
47(2), 788.</p></li>
</ul>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Francisco Pascoal, Paula Branco, Luís Torgo, Rodrigo Costa, Catarina Magalhães.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
